### PR TITLE
Support for plasticity

### PR DESCRIPTION
**Pull request by boriskaus**
This PR adds support for plasticity as part of the `CompositeRheology` structure. 

A simple viscoelastoplastic setup:
```Julia
julia> v1 = SetDiffusionCreep("Dry Anorthite | Rybacki et al. (2006)");
julia> v2 = SetDislocationCreep("Dry Anorthite | Rybacki et al. (2006)");
julia> v3 = LinearViscous();
julia> pl1= DruckerPrager(C=1e6); 
julia> e1 = ConstantElasticity();   
julia> c1 = CompositeRheology(v1,v2,e1,pl1)
--⟦▪̲̅▫̲̅▫̲̅▫̲̅----⟦▪̲̅▫̲̅▫̲̅▫̲̅----/\/\/----▬▬▬__--

julia> args = (T=900.0, d=100e-6, τII_old=7e5, dt=8e8, P=0.0);
julia> τ    = compute_τII(c1, εII, args, verbose=true)
τII guess = 939311.2212556291
 iter 1 0.08462502224446022 F=-73285.81747119047
 iter 2 4.178154614552779e-17 F=0.0
---
866025.4037844386
```

Doing the same with a (regularised) visco-elasto-viscoplastic rheology:
```Julia
julia> c2 = CompositeRheology(v1,v2,e1,Parallel(pl1,v3))
--⟦▪̲̅▫̲̅▫̲̅▫̲̅----⟦▪̲̅▫̲̅▫̲̅▫̲̅----/\/\/--|--▬▬▬__--|
                          |--⟦▪̲̅▫̲̅▫̲̅▫̲̅--|

julia> τ    = compute_τII(c2, εII, args, verbose=true)
τII guess = 939311.2212556291
 iter 1 0.10687619449000424 F=-73285.81747119047
 iter 2 1.0304754152958193e-13 F=0.0
---
918864.2018406295     
```

NOTE: this does not yet work in combination with automatic differentiation, that is `compute_τII_AD` and friends are not yet implemented.

### CODE REVIEW

File: src/CompositeRheologies.jl [hunk 1]
```diff
@@ -557,6 +716,90 @@ end
...
+        return nothing
+    end
+end
+
+@inline function _fill_J_plastic!(J, r, x, element, args, is_plast, is_par, j)
+    !is_plast && return j
+
+    j       += 1
+    λ̇       = x[j]
+    id_par  = findall(is_par);
```
Comment (albert-de-montserrat) [inline src/CompositeRheologies.jl:733]: I'm confused by this line, isn't `ispar` just `true` or `false`?

Comment (boriskaus) [reply src/CompositeRheologies.jl]: yes, `ispar` is indeed `true/false`; how do you suggest to do it instead?

File: src/CompositeRheologies.jl [hunk 2]
```diff
@@ -35,28 +46,45 @@ end
...
     Npar        =   length(id_parallel)
     n           =   length(v)
     par         =   zeros(Bool,n)
     if Npar>0
         par[id_parallel] .= 1
     end
-    is_parallel =    SVector{n,Bool}(par)
-
-    return CompositeRheology{typeof(v), n, Npar, is_parallel, τ_it, P_it, λ_it}(v)
+    is_parallel =   SVector{n,Bool}(par)
```
[the remark this answers is not in the document]
Comment (boriskaus) [reply src/CompositeRheologies.jl:71]: I suppose that a tuple would do as well.

[the remark this answers is not in the document]
Comment (boriskaus) [reply src/CompositeRheologies.jl]: changed that, which simplifies the code

File: src/CompositeRheologies.jl (hunk 1 above)
Comment (boriskaus) [reply src/CompositeRheologies.jl]: was changed meanwhile

File: src/CompositeRheologies.jl [hunk 3]
```diff
@@ -527,13 +578,121 @@ This performs nonlinear Newton iterations for `τII` with given `εII_total` for
         error("iterations did not converge")
     end
 
+    if full_output
```
[the remark this answers is not in the document]
Comment (boriskaus) [reply src/CompositeRheologies.jl]: added that

File: src/CompositeRheologies.jl [hunk 4]
```diff
@@ -529,13 +558,106 @@ This performs nonlinear Newton iterations for `τII` with given `εII_total` for
...
+
+    end
+
+    r = @MVector zeros(_T,n);
+    J = @MMatrix zeros(_T, n,n)   # size depends on # of plastic elements
+    
+    # Local Iterations
+    iter = 0
+    ϵ = 2 * tol
+    τII_prev = τ_initial
```
Comment (albert-de-montserrat) [inline src/CompositeRheologies.jl:627]: `τII_prev` and `τ_parallel` seemed to be unused

File: src/CompositeRheologies.jl [hunk 5]
```diff
@@ -729,6 +1023,7 @@ end
 
 _compute_εII_nonparallel(v, TauII::_T, args) where {_T} = compute_εII(v, TauII, args)
 _compute_εII_nonparallel(v::Parallel, TauII::_T, args) where {_T} = zero(_T)
+_compute_εII_nonparallel(v::AbstractPlasticity, TauII::_T, args) where {_T} = zero(_T)
 
 
 # 0D RHEOLOGY functions
```
[the remark this answers is not in the document]
Comment (boriskaus) [reply src/CompositeRheologies.jl]: good point; done that

File: src/CompositeRheologies.jl (hunk 4 above)
Comment (boriskaus) [reply src/CompositeRheologies.jl]: removed

File: src/CompositeRheologies.jl [hunk 6]
```diff
@@ -529,13 +558,104 @@ This performs nonlinear Newton iterations for `τII` with given `εII_total` for
...
+    x[1] = τ_initial
+
+    j = 1;
+    for i=1:N
+        #if is_par[i]
+        #    j += 1
+        #    x[j] = compute_εII_harmonic_i(c, τ_initial, args,i)   
+        #end
+
+        if is_plastic[i] & is_par[i]
```
Comment (albert-de-montserrat) [inline src/CompositeRheologies.jl:605]: Should be the short-circuit operator `&&`, same below

File: src/CompositeRheologies.jl [hunk 7]
```diff
@@ -550,15 +670,103 @@ end
...
+        τ       = x[1]
+        τ_pl    = x[j+1]    # if the plastic element is in || with other elements, need to explicitly solve for this
+
+        args    = merge(args, (τII=τ_pl,))
+        F       = compute_yieldfunction(element,args);  # yield function applied to plastic element
+    
+        ε̇_pl    =  λ̇*∂Q∂τII(element, τ_pl)  
+        r[1]   -=  ε̇_pl                     #  add plastic strainrate
+        
+        if F>0
```
Comment (albert-de-montserrat) [inline src/CompositeRheologies.jl:708]: `F > 0.0`, same below

File: src/CompositeRheologies.jl (hunk 6 above)
Comment (boriskaus) [reply src/CompositeRheologies.jl]: Done

File: src/CompositeRheologies.jl (hunk 7 above)
Comment (boriskaus) [reply src/CompositeRheologies.jl]: Implemented

**Review by albert-de-montserrat — APPROVED** [review]
Happy with all the code changes!  From my side, feel free to merge.

**Pull request merged**